src/engine/filters/index.ts

```diff
@@ -0,0 +1,28 @@
+import {
+  AudienceDefinitionFilter,
+  EngineConditionQuery,
+  PageFeatureResult,
+} from '../../../types';
+import {
+  versionMatches,
+  arrayIntersectsCondition,
+  vectorDistanceCondition,
+  cosineSimilarityCondition,
+} from './conditions';
+
+export const matchesQuery = (
+  query: EngineConditionQuery<AudienceDefinitionFilter>,
+  queryFeatures: PageFeatureResult
+): boolean => {
+  // Can queryFeatures be empty?
+  if (!queryFeatures || !versionMatches(queryFeatures, query)) {
+    return false;
+  }
+
+  // matches if any of the conditions are satisfied
+  return [
+    arrayIntersectsCondition,
+    vectorDistanceCondition,
+    cosineSimilarityCondition,
+  ].some((match) => match(queryFeatures, query));
```

Comment:
This feels a little inefficient to just run all the conditions on all the page views?

Reply:
I think the intersection / vector distances are not actually computed once the type guard mismatch.
In the case where the VM optimizes the early returning on the checks and over the disjunction (some), it should be actually equivalent to the ternary type checking.
But I can be wrong... and I'm also not sure that it optimizes the disjunction...
I'll write a benchmark later to check if it blocks the client for too much time.